llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -753,36 +753,302 @@ DebugAbbrevSectionRef::materialize(MCCASReader &Reader,
+    // when materializing and creating the CAS.
+    IsSetFile = true;
+  } break;
+  case dwarf::DW_LNS_fixed_advance_pc: {
+    // Takes a single uhalf operand, move cursor to the end of operand.
+    LineTableDataReader.getRelocatedValue(LineTableCursor, 2);
+  } break;
+  default:
+    llvm_unreachable("Unknown standard opcode for line table");
+    break;
+  }
+  return Error::success();
+}
+
+static Expected<std::pair<uint64_t, uint64_t>>
+getOpcodeAndOperandSize(StringRef DistinctData, StringRef LineTableData,
+                        uint64_t DistinctOffset, uint64_t LineTableOffset,
+                        bool IsLittleEndian, uint8_t OpcodeBase) {
+  DWARFDataExtractor LineTableDataReader(LineTableData, IsLittleEndian, 8);
+  DWARFDataExtractor DistinctDataReader(DistinctData, IsLittleEndian, 8);
+  DWARFDataExtractor::Cursor LineTableCursor(LineTableOffset);
+  DWARFDataExtractor::Cursor DistinctCursor(DistinctOffset);
+
+  auto Opcode = LineTableDataReader.getU8(LineTableCursor);
+  if (Opcode == 0) {
+    // Extended Opcodes always start with a zero opcode followed by
+    // a uleb128 length so you can skip ones you don't know about
+    uint64_t Len = LineTableDataReader.getULEB128(LineTableCursor);
+    if (Len == 0)
+      return createStringError(inconvertibleErrorCode(),
+                               "0 Length for an extended opcode is wrong");
+
+    uint8_t SubOpcode = LineTableDataReader.getU8(LineTableCursor);
+    bool IsEndSequence = false;
+    auto Err = handleExtendedOpcodesForLineTable(
+        LineTableDataReader, LineTableCursor, SubOpcode, Len, IsEndSequence);
+    if (Err)
+      return std::move(Err);
+    if (IsEndSequence) {
+      // The SubOpcode is a DW_LNE_end_sequence, it takes no operand, but check
+      // if this is the end of the line table and return.
+      assert(LineTableData.size() == LineTableCursor.tell() &&
+             "Malformed Line Table, data exists after a DW_LNE_end_sequence");
+    }
+  } else if (Opcode < OpcodeBase) {
+    bool IsSetFile = false;
+    auto Err = handleStandardOpcodesForLineTable(
+        LineTableDataReader, LineTableCursor, Opcode, IsSetFile);
+    if (Err)
+      return std::move(Err);
+    if (IsSetFile) {
+      // The Opcode is DW_LNS_set_file, this means we need to get the file
+      // number from the DistinctData, which is stored as a ULEB.
+      DistinctDataReader.getULEB128(DistinctCursor);
+    }
+  } else {
+    // Special Opcodes, do nothing.
+  }
+
+  if (!LineTableCursor)
+    return LineTableCursor.takeError();
+  if (!DistinctCursor)
+    return DistinctCursor.takeError();
+
+  return std::make_pair(LineTableCursor.tell() - LineTableOffset,
+                        DistinctCursor.tell() - DistinctOffset);
+}
+
+static Expected<SmallVector<char, 0>>
+materializeDebugLineSection(MCCASReader &Reader,
+                            const SmallVector<cas::ObjectRef> &Refs) {
```

Comment:
array ref

Reply:
Ugh, I always forget that ArrayRef exists! Good catch!